File: app/src/main/java/com/fastaccess/ui/modules/repos/extras/milestone/create/CreateMilestoneDialogFragment.java
package com.fastaccess.ui.modules.repos.extras.milestone.create;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.TextInputEditText;
import android.support.design.widget.TextInputLayout;
import android.support.v7.widget.Toolbar;
import android.view.MotionEvent;
import android.view.View;

import com.fastaccess.R;
import com.fastaccess.data.dao.MilestoneModel;
import com.fastaccess.datetimepicker.DatePickerFragmentDialog;
import com.fastaccess.datetimepicker.callback.DatePickerCallback;
import com.fastaccess.helper.BundleConstant;
import com.fastaccess.helper.Bundler;
import com.fastaccess.helper.InputHelper;
import com.fastaccess.helper.ParseDateFormat;
import com.fastaccess.ui.base.BaseDialogFragment;

import butterknife.BindView;
import butterknife.OnTouch;

/**
 * Created by Kosh on 04 Mar 2017, 10:40 PM
 */

public class CreateMilestoneDialogFragment extends BaseDialogFragment<CreateMilestoneMvp.View, CreateMilestonePresenter>
        implements CreateMilestoneMvp.View, DatePickerCallback {

    public static final String TAG = CreateMilestoneDialogFragment.class.getSimpleName();

    @BindView(R.id.toolbar) Toolbar toolbar;
    @BindView(R.id.title) TextInputLayout title;
    @BindView(R.id.dueOnEditText) TextInputEditText dueOnEditText;
    @BindView(R.id.description) TextInputLayout description;

    private CreateMilestoneMvp.OnMilestoneAdded onMilestoneAdded;

    public static CreateMilestoneDialogFragment newInstance(@NonNull String login, @NonNull String repo) {
        CreateMilestoneDialogFragment fragment = new CreateMilestoneDialogFragment();
        fragment.setArguments(Bundler.start()
                .put(BundleConstant.EXTRA, login)
                .put(BundleConstant.ID, repo)
                .end());
        return fragment;
    }

    @Override public void onAttach(Context context) {
        super.onAttach(context);
        if (getParentFragment() instanceof CreateMilestoneMvp.OnMilestoneAdded) {
            onMilestoneAdded = (CreateMilestoneMvp.OnMilestoneAdded) getParentFragment();
        } else {
            onMilestoneAdded = (CreateMilestoneMvp.OnMilestoneAdded) context;
        }
    }

    @Override public void onDetach() {
        onMilestoneAdded = null;
        super.onDetach();
    }

    @Override protected int fragmentLayout() {
        return R.layout.create_milestone_layout;
    }

    @Override protected void onFragmentCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        String login = getArguments().getString(BundleConstant.EXTRA);
        String repo = getArguments().getString(BundleConstant.ID);
        if (login == null || repo == null) {
            return;
        }
        toolbar.setTitle(R.string.create_milestone);
        toolbar.setNavigationIcon(R.drawable.ic_clear);
        toolbar.setNavigationOnClickListener(item -> dismiss());
        toolbar.inflateMenu(R.menu.add_menu);
        toolbar.getMenu().findItem(R.id.add).setIcon(R.drawable.ic_send);
        toolbar.setOnMenuItemClickListener(item -> {
            getPresenter().onSubmit(InputHelper.toString(title), InputHelper.toString(dueOnEditText), InputHelper.toString(description), login, repo);
            return true;
        });
    }

    @NonNull @Override public CreateMilestonePresenter providePresenter() {
        return new CreateMilestonePresenter();
    }

    @OnTouch(R.id.dueOnEditText) boolean onTouch(MotionEvent event) {
        if (event.getAction() == MotionEvent.ACTION_UP) {
            DatePickerFragmentDialog.newInstance().show(getChildFragmentManager(), "DatePickerFragmentDialog");
        }
        return false;
    }

    @Override public void onDateSet(long date) {
        if (date > 0) {
            dueOnEditText.setText(ParseDateFormat.prettifyDate(date));
        }
    }

    @Override public void onShowTitleError(boolean isError) {
        title.setError(isError ? getString(R.string.required_field) : null);
    }

    @Override public void onMilestoneAdded(@NonNull MilestoneModel milestoneModel) {
        hideProgress();
        onMilestoneAdded.onMilestoneAdded(milestoneModel);
        dismiss();
    }
}

File: app/src/main/java/com/fastaccess/github/ui/modules/issuesprs/edit/milestone/CreateMilestoneDialogFragment.kt
package com.fastaccess.github.ui.modules.issuesprs.edit.milestone

import android.content.Context
import android.os.Bundle
import android.view.MotionEvent
import android.view.View
import com.fastaccess.datetimepicker.DatePickerFragmentDialog
import com.fastaccess.datetimepicker.callback.DatePickerCallback
import com.fastaccess.github.R
import com.fastaccess.github.base.BaseDialogFragment
import com.fastaccess.github.extensions.DatePrettifier
import com.fastaccess.github.utils.extensions.asString
import kotlinx.android.synthetic.main.add_milestone_layout.*
import java.util.*

/**
 * Created by Kosh on 30.03.19.
 */
class CreateMilestoneDialogFragment : BaseDialogFragment(), DatePickerCallback {
    private var callback: OnAddNewMilestone? = null

    override fun layoutRes(): Int = R.layout.add_milestone_layout

    override fun onAttach(context: Context) {
        super.onAttach(context)
        callback = when {
            parentFragment is OnAddNewMilestone -> parentFragment as OnAddNewMilestone
            context is OnAddNewMilestone -> context
            else -> throw IllegalAccessException("woops, your $context or $parentFragment must impl OnAddNewMilestone")
        }
    }

    override fun onDetach() {
        callback = null
        super.onDetach()
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setupToolbar(getString(R.string.create_milestone), R.menu.submit_menu) {
            if (it.itemId == R.id.submit) {
                val timestamp = dueOn.tag as? Long
                val title = milestoneTitle.editText?.asString()
                val description = milestoneDescription.editText?.asString()
                milestoneTitle.error = if (title.isNullOrEmpty()) getString(R.string.required_field) else null
                dueOn.error = if (timestamp == null) getString(R.string.required_field) else null
                if (timestamp != null && !title.isNullOrEmpty()) {
                    callback?.addNewMilestone(title, Date(timestamp), description)
                    dismissDialog()
                }
            }
        }
        dueOn.editText?.setOnTouchListener { _, event ->
            if (event.action == MotionEvent.ACTION_UP) {
                DatePickerFragmentDialog.newInstance().show(childFragmentManager, "DatePickerFragmentDialog")
                return@setOnTouchListener true
            }
            return@setOnTouchListener false
        }
    }

    override fun onDateSet(date: Long) {
        dueOn.tag = date
        dueOn.editText?.setText(DatePrettifier.prettifyDate(date))
    }

    interface OnAddNewMilestone {
        fun addNewMilestone(title: String, dueOn: Date, description: String?)
    }
}